kuma/core/middleware.py

```diff
@@ -8,82 +7,204 @@
 from django.http import (HttpResponseForbidden,
                          HttpResponsePermanentRedirect,
                          HttpResponseRedirect)
+from django.middleware.locale import (
+    LocaleMiddleware as DjangoLocaleMiddleware)
 from django.utils import translation
 from django.utils.encoding import iri_to_uri, smart_str
+from django.utils.six.moves.urllib.parse import (
+    urlencode, urlsplit, urlunsplit)
 from whitenoise.middleware import WhiteNoiseMiddleware
 
 from .decorators import add_shared_cache_control
-from .urlresolvers import Prefixer, set_url_prefixer, split_path
-from .utils import is_untrusted, urlparams
+from .i18n import django_language_code_to_kuma
+from .utils import is_untrusted
 from .views import handler403
 
 
-class LocaleURLMiddleware(object):
+class LangSelectorMiddleware(object):
     """
-    Based on zamboni.amo.middleware.
-    Tried to use localeurl but it choked on 'en-US' with capital letters.
+    Redirect requests with a ?lang= parameter.
 
-    1. Search for the locale.
-    2. Save it in the request.
-    3. Strip them from the URL.
+    This should appear higher than LocaleMiddleware in the middleware list.
     """
 
     def process_request(self, request):
-        prefixer = Prefixer(request)
-        set_url_prefixer(prefixer)
-        full_path = prefixer.fix(prefixer.shortened_path)
-        lang = request.GET.get('lang')
-
-        if lang in dict(settings.LANGUAGES):
-            # Blank out the locale so that we can set a new one. Remove lang
-            # from the query params so we don't have an infinite loop.
-            prefixer.locale = ''
-            new_path = prefixer.fix(prefixer.shortened_path)
-            query = dict((smart_str(k), v) for
+        """Redirect if ?lang query parameter is valid."""
+        query_lang = request.GET.get('lang')
+        if query_lang not in dict(settings.LANGUAGES):
+            # Invalid language requested, don't redirect
+            return
+
+        # Check if different language is embedded in URL
+        dj_language = translation.get_language_from_request(
+            request, check_path=True)
+        language = django_language_code_to_kuma(dj_language)
```

Comment:
From my tests and reading of the Django code, it seems to me that `translation.get_language_from_request` will always return the Kuma version of the language code, `en-US` instead of `en-us`, since the Django code uses `settings.LANGUAGES`, which contains the Kuma language codes. So the call to `django_language_code_to_kuma` doesn't seem necessary, unless it's not necessary now but will be in the context of future updates for the Django 1.11 upgrade?